packages/protocol/src/extension.js

```diff
@@ -0,0 +1,310 @@
+   */
+  _pendingMessages = new Map();
+
+  /**
+   * Handshake handler.
+   * @type {Function<{protocol, context}>}
+   */
+  _handshakeHandler = null;
+
+  /**
+   * Close handler.
+   * @type {Function<{protocol, context}>}
+   */
+  _closeHandler = null;
+
+  /**
+   * Message handler.
+   * @type {Function<{protocol, context, message}>}
+   */
+  _messageHandler = null;
+
+  /**
+   * Feed handler.
+   * @type {Function<{protocol, context, discoveryKey}>}
+   */
+  _feedHandler = null;
+
+  _stats = {
+    send: 0,
+    receive: 0,
+    error: 0
+  };
+
+  /**
+   * @param {string} name
+   * @param {Object} options
+   * @param {Number} options.timeout
+   * @param {Codec} options.codec
+   */
+  constructor (name, options = {}) {
+    super();
+    assert(typeof name === 'string' && name.length > 0, 'Name is required.');
+
+    this._name = name;
+
+    this._options = Object.assign({
+      timeout: 2000,
+      codec: new Codec({ binary: options.binary })
+    }, options);
+  }
+
+  get name () {
+    return this._name;
+  }
+
+  get stats () {
+    return this._stats;
+  }
+
+  /**
+   * Sets the handshake handler.
+   * @param {Function<{protocol, context}>} handshakeHandler - Async handshake handler.
+   * @returns {Extension}
+   */
+  setHandshakeHandler (handshakeHandler) {
+    this._handshakeHandler = handshakeHandler;
+
+    return this;
+  }
+
+  /**
+   * Sets the close stream handler.
+   * @param {Function<{protocol, context}>} closeHandler - Close handler.
+   * @returns {Extension}
+   */
+  setCloseHandler (closeHandler) {
+    this._closeHandler = closeHandler;
+
+    return this;
+  }
+
+  /**
+   * Sets the message handler.
+   * @param {Function<{protocol, context, message}>} messageHandler - Async message handler.
+   * @returns {Extension}
+   */
+  setMessageHandler (messageHandler) {
+    this._messageHandler = messageHandler;
+
+    return this;
+  }
+
+  /**
+   * Sets the message handler.
+   * @param {Function<{protocol, context, discoveryKey}>} feedHandler - Async feed handler.
+   * @returns {Extension}
+   */
+  setFeedHandler (feedHandler) {
+    this._feedHandler = feedHandler;
+
+    return this;
+  }
+
+  /**
+   * Initializes the extension.
+   *
+   * @param {Protocol} protocol
+   */
+  init (protocol) {
+    assert(!this._protocol);
+    log(`init[${this._name}]: ${keyToHuman(protocol.id)}`);
+
+    this._protocol = protocol;
+  }
+
+  /**
+   * Handshake event.
+   *
+   * @param {Object} context
+   */
+  onHandshake (context) {
+    if (this._handshakeHandler) {
+      this._handshakeHandler(this._protocol, context);
+    }
+  }
+
+  /**
+   * Close event.
+   *
+   * @param {Object} context
+   */
+  onClose (error, context) {
+    if (this._closeHandler) {
+      this._closeHandler(error, this._protocol, context);
+    }
+  }
+
+  /**
+   * Receives extension message.
+   *
+   * @param {Object} context
+   * @param {Buffer} message
+   */
+  async onMessage (context, message) {
```

Comment:
Should this be async (does the caller wait for this?) If so please add @return type.